src/imghash.rs

```diff
@@ -1,37 +1,104 @@
+            panic!("Matrix cannot be empty");
+        }
+
         // Ensures that the matrix rows all have the same length.
         // If not, this is a critical issue and likely a bug in the code
         // that creates the hash -> therefore a panic here is appropriate
-        let first_length = matrix.first().unwrap().len();
-        matrix.iter().for_each(|row| {
-            if row.len() != first_length {
-                panic!("All rows must have the same length");
-            }
+        let width = matrix.first().unwrap().len();
+        let height = matrix.len();
+
+        if matrix.iter().any(|row| row.len() != width) {
+            panic!("All rows must have the same length");
+        }
+
+        Self::from_stream(
+            matrix.into_iter().flat_map(Vec::into_iter),
+            width as u32,
+            height as u32,
+        )
+    }
+
+    /// Create a new ImageHash from the specified bits stream.
+    pub(crate) fn from_stream(
+        stream: impl IntoIterator<Item = bool>,
+        width: u32,
+        height: u32,
+    ) -> ImageHash {
+        let length = width as usize * height as usize;
+        let size = (length + 7) / 8;
+        let padding = size * 8 - length;
+
+        let buf = vec![0_u8; size];
+        let mut data = BitBox::from_bitslice(&buf.view_bits::<Msb0>()[padding..]);
+        let mut count = 0;
+
+        stream.into_iter().enumerate().for_each(|(i, bit)| {
+            data.set(i, bit);
+            count += 1;
         });
 
-        ImageHash { matrix }
+        if count != length {
+            panic!("Data length does not match the specified width and height");
+        }
+
+        data.fill_uninitialized(false);
+
+        ImageHash {
+            data,
+            width,
+            height,
+        }
     }
 
     /// Returns a copy of the underlying matrix that represents the [`ImageHash`].
+    #[deprecated(
+        since = "1.5.0",
+        note = "This method is inefficient because it creates a new Vec<Vec<bool>>."
+    )]
     pub fn matrix(&self) -> Vec<Vec<bool>> {
-        self.matrix.clone()
+        let width = self.width as usize;
+
+        (0..self.height as usize)
+            .map(|y| {
+                (0..self.width as usize)
+                    .map(|x| {
+                        self.data
+                            .get(y * width + x)
+                            .as_deref()
+                            .copied()
+                            .unwrap_or(false)
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>()
     }
 
     /// Flattens the bit matrix that represents the [`ImageHash`] into a single vector.
+    #[deprecated(
+        since = "1.5.0",
+        note = "This method is inefficient because it creates a new Vec<bool>."
+    )]
     pub fn flatten(&self) -> Vec<bool> {
-        self.matrix.iter().flatten().copied().collect()
+        self.data.iter().map(|v| *v).collect()
```

Comment:
Meh even if both works just fine, it is probably a little cleaner to use the provided `by_vals()`.

```suggestion
        self.data.iter().by_vals().collect()
```